src/DebugEngineHost/HostLogger.cs

```diff
@@ -3,101 +3,151 @@
+        Error,
+        /// <summary>
+        /// Logs that describe an unrecoverable application or system crash, or a catastrophic failure that requires immediate attention.
+        /// </summary>
+        Critical,
+        /// <summary>
+        /// Not used for writing log messages.
+        /// Specifies that a logging category should not write any messages.
         /// </summary>
-        /// <param name="outputString"></param>
-        public delegate void OutputCallback(string outputString);
+        None
+    }
 
-        private StreamWriter _streamWriter;
-        private OutputCallback _callback;
-        private readonly object _locker = new object();
+    public class HostLogChannel
+    {
+        private readonly Action<LogLevel, string> _log;
+        private readonly StreamWriter _logFile;
 
-        internal HostLogger(StreamWriter streamWriter = null, OutputCallback callback = null)
-        {
-            _streamWriter = streamWriter;
-            _callback = callback;
-        }
+        private readonly object _lock = new object();
 
-        public void WriteLine(string line)
+        public HostLogChannel(Action<LogLevel, string> logAction, string file)
         {
-            lock (_locker)
+            _log = logAction;
+
+            if (!string.IsNullOrEmpty(file))
             {
-                if (_streamWriter != null)
-                    _streamWriter.WriteLine(line);
-                _callback?.Invoke(line);
+                _logFile = File.CreateText(file);
             }
         }
 
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="verbosity"></param>
+        /// <param name="message"></param>
+        public void WriteLine(LogLevel verbosity, string message)
+        {
+            _log?.Invoke(verbosity, message);
+            _logFile?.WriteLine(message);
+            _logFile?.Flush();
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="verbosity"></param>
+        /// <param name="format"></param>
+        /// <param name="values"></param>
+        public void WriteLine(LogLevel verbosity, string format, params object[] values)
+        {
+            string message = string.Format(CultureInfo.InvariantCulture, format, values);
+            _log?.Invoke(verbosity, message);
+            _logFile?.WriteLine(message);
+            _logFile?.Flush();
+        }
+
         public void Flush()
         {
-            lock (_locker)
-            {
-                if (_streamWriter != null)
-                    _streamWriter.Flush();
-            }
+            _logFile?.Flush();
         }
 
         public void Close()
         {
-            lock (_locker)
-            {
-                if (_streamWriter != null)
-                    _streamWriter.Close();
-                _streamWriter = null;
-            }
+            _logFile?.Close();
         }
+    }
+
+    public sealed class HostLogger
```

Comment:
>HostLogger [](http://example.com/codeflow?start=24&length=10)

Feels like somewhere in here we should be reading the VS registry to see what the natvis setting should be, and maybe setting up a write watcher to update it if it changes.